File: blender_extension/roblox_interface/roblox_server/format_types/register_mask.py
def read(self, format_data, **masks):
	register_masks = format_data["register_mask"]
	format = format_data["count_format"]
	bitmask, = format.unpack_from(*self.__read_buffer(format.size))

	bools = []
	for _ in range(len(register_masks)):
		bools.insert(0, bool(bitmask & 0b1))
		bitmask >>= 1
		
	for mask, bool_value in zip(register_masks, bools):
		masks[mask] = bool_value

	args = tuple(bools)
	args += self.__parse(format_data["data"], **masks)
	return args

def write(self, args, args_count, format_data, **masks):
	register_masks = format_data["register_mask"]
	mask_count = len(register_masks)
	bools = args[args_count:args_count + mask_count]
	args_count += mask_count

	bitmask = 0
	for bool_value in bools:
		bitmask <<= 1
		bitmask += 1 if bool_value else 0
	
	for mask, bool_value in zip(register_masks, bools):
		masks[mask] = bool_value

	format = format_data["count_format"]
	self.__write_buffer(format.pack(bitmask), format.size)
	args_count = self.__parse(args, args_count, format_data["data"], **masks)
	return args_count

File: roblox_plugin/src/Server/formatTypes/register_mask.ts
//!native
import * as ProcessFormats from "../processFormats";
import SendMessagesThread from "../sendMessages";
import ReceiveMessagesThread from "../receiveMessages";

interface RegisterMaskFormatData {
	type: "register_mask";
	register_mask: string[];
	data: ProcessFormats.FormatData;
}
export function read(
	receiveThread: ReceiveMessagesThread,
	args: defined[],
	formatData: RegisterMaskFormatData,
	masks: Map<string, boolean>,
) {
	const registerMasks = formatData.register_mask;
	const maskCount = registerMasks.size();
	const countFormat = ProcessFormats.getFormatForCount(maskCount);

	const [readBuffer, readOffset] = ReceiveMessagesThread.readBuffer(receiveThread, countFormat.size);
	let bitmask = countFormat.read(readBuffer, readOffset);
	const bools: boolean[] = [];
	for (const _ of $range(0, maskCount - 1)) {
		bools.insert(0, !!(bitmask & 1));
		bitmask >>= 1;
	}

	const newMasks = ProcessFormats.deepcopy(masks);
	for (const i of $range(0, maskCount - 1)) {
		args.push(bools[i]);
		newMasks.set(registerMasks[i], bools[i]);
	}
	ReceiveMessagesThread.parse(receiveThread, args, formatData.data, newMasks);
}
export function write(
	sendThread: SendMessagesThread,
	args: defined[],
	argsCount: number,
	formatData: RegisterMaskFormatData,
	masks: Map<string, boolean>,
): number {
	const registerMasks = formatData.register_mask;
	const maskCount = registerMasks.size();
	const countFormat = ProcessFormats.getFormatForCount(maskCount);

	let bitmask = 0;
	const newMasks = ProcessFormats.deepcopy(masks);
	for (const i of $range(0, maskCount - 1)) {
		const boolValue = args[argsCount] as boolean;
		argsCount += 1;
		bitmask <<= 1;
		bitmask += boolValue ? 1 : 0;
		newMasks.set(registerMasks[i], boolValue);
	}
	const data = buffer.create(countFormat.size);
	countFormat.write(data, 0, bitmask);
	SendMessagesThread.writeBuffer(sendThread, data, countFormat.size);
	return SendMessagesThread.parse(sendThread, args, argsCount, formatData.data, newMasks);
}
